src/rules.jl

```diff
@@ -1,260 +1,313 @@
+            if phase(zxd, v1) == 0 && length(neighbors(zxd, v1, count_mul = true)) == 2
+                push!(matches, Match{T}([v1], Rule_i1()))
+            end
         end
-    else
-        msg = "Spider $(v) is not in this ZX-diagram!"
     end
-    return msg
+    return matches
 end
 
-function rule_h!(zxd::ZXDiagram{T,U,P}, v::T) where {T<:Integer, U<:Integer, P}
-    msg = check_rule_h(zxd, v)
-    if msg != true
-        error(msg)
-    end
-    zxd.st[v] = SType(1 - Int(zxd.st[v]))
-    for v1 in outneighbors(zxd, v)
-        insert_spider!(zxd, v1, v, H)
+function match_rule(::Rule_i2, zxd::ZXDiagram{T, P}) where {T, P}
+    matches = Match{T}[]
+    for v1 in spiders(zxd)
+        if spider_type(zxd, v1) == H && length(neighbors(zxd, v1, count_mul = true)) == 2
+            for v2 in neighbors(zxd, v1)
+                if spider_type(zxd, v2) == H && length(neighbors(zxd, v2, count_mul = true)) == 2
+                    v2 >= v1 && push!(matches, Match{T}([v1, v2], Rule_i2()))
+                end
+            end
+        end
     end
-    zxd
+    return matches
 end
 
-function check_rule_i1(zxd::ZXDiagram{T,U,P}, v::T) where {T<:Integer, U<:Integer, P}
-    msg = true
-    if v <= nv(zxd)
-        if zxd.st[v] == Z || zxd.st[v] == X
-             if zxd.ps[v] == zero(P)
-                 if degree(zxd.g, v) != 2
-                     msg = "Spider $(v) is not of degree 2!"
-                 end
-             else
-                 msg = "The phase of spider $(v) is not 0!"
-             end
-        else
-            msg = "Spider $(v) is not a Z or X spider!"
+function match_rule(::Rule_pi, zxd::ZXDiagram{T, P}) where {T, P}
+    matches = Match{T}[]
+    for v1 in spiders(zxd)
+        if spider_type(zxd, v1) == X && phase(zxd, v1) == one(P) && length(neighbors(zxd, v1, count_mul = true)) == 2
+            for v2 in neighbors(zxd, v1)
+                if spider_type(zxd, v2) == Z
+                    push!(matches, Match{T}([v1, v2], Rule_pi()))
+                end
+            end
         end
-    else
-        msg = "Spider $(v) is not in this ZX-diagram!"
     end
-    return msg
+    return matches
 end
 
-function rule_i1!(zxd::ZXDiagram{T,U,P}, v::T) where {T<:Integer, U<:Integer, P}
-    msg = check_rule_i1(zxd, v)
-    if msg != true
-        error(msg)
+function match_rule(::Rule_c, zxd::ZXDiagram{T, P}) where {T, P}
+    matches = Match{T}[]
+    for v1 in spiders(zxd)
+        if spider_type(zxd, v1) == X && phase(zxd, v1) == zero(P) && length(neighbors(zxd, v1, count_mul = true)) == 1
+            for v2 in neighbors(zxd, v1)
+                if spider_type(zxd, v2) == Z
+                    push!(matches, Match{T}([v1, v2], Rule_c()))
+                end
+            end
+        end
     end
-    nbhd = outneighbors(zxd, v)
-    if length(nbhd) == 2
-        add_edge!(zxd.g, nbhd[1], nbhd[2])
+    return matches
+end
+
+function match_rule(::Rule_b, zxd::ZXDiagram{T, P}) where {T, P}
```

Comment:
just use `match` instead of `match_rule`? I don't see any ambiguity here.

Reply:
what we are doing here is actually exactly pattern matching on graphs, so since we might integrate with `SymbolicUtils` in the future to use their symbolic support on phases and mixed with graph pattern match (see discussion here: https://github.com/JuliaSymbolics/SymbolicUtils.jl/issues/89 )